atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/id/SweepTableIndices.java

```diff
@@ -0,0 +1,104 @@
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.logsafe.SafeArg;
+
+/**
+ * Creates a dictionary of table references to shorter (integral) identifiers.
+ * <p>
+ * Algorithm is slightly more complicated than usual Atlas code because it cannot assume transactions. It works as
+ * follows.
+ * <p>
+ * 1. If table exists in NamesToIds and is 'identified', we are done.
+ * 2. If table is 'pending', take its id. Otherwise PutUnlessExists in the current highest known id + 1 in IdsToNames
+ *    into the NamesToIds table. In case of failure, take the value that now must exist (and if now 'identified', we
+ *    are done).
+ * 3. PutUnlessExists that value into IdsToNames. If success or it already existed with the same value, go to 4,
+ *    otherwise 5.
+ * 4. CAS NamesToIds from 'pending' to 'identified'.
+ * 5. CAS NamesToIds from its current value to the current highest known id + 1; the current best candidate has been
+ *    already used by a concurrent writer.
+ * <p>
+ */
+public final class SweepTableIndices {
+    private static final Logger log = LoggerFactory.getLogger(SweepTableIndices.class);
+
+    private final IdsToNames idToNames;
+    private final NamesToIds namesToIds;
+    private final LoadingCache<TableReference, Integer> tableIndices;
+    private final LoadingCache<Integer, TableReference> tableRefs;
+
+    SweepTableIndices(IdsToNames idsToNames, NamesToIds namesToIds) {
+        this.idToNames = idsToNames;
+        this.namesToIds = namesToIds;
+        this.tableIndices = Caffeine.newBuilder().maximumSize(20_000).build(this::loadUncached);
+        this.tableRefs = Caffeine.newBuilder().maximumSize(20_000).build(this::getTableReferenceUncached);
+    }
+
+    public SweepTableIndices(KeyValueService kvs) {
+        this(new IdsToNames(kvs), new NamesToIds(kvs));
+    }
+
+    public int getTableId(TableReference table) {
+        return tableIndices.get(table);
+    }
+
+    public TableReference getTableReference(int tableId) {
+        return tableRefs.get(tableId);
+    }
+
+    private TableReference getTableReferenceUncached(int tableId) {
+        return idToNames.get(tableId)
+                .orElseThrow(() -> new NoSuchElementException("Id " + tableId + " does not exist"));
+    }
+
+    private int loadUncached(TableReference table) {
+        while (true) {
+            Optional<SweepTableIdentifier> identifier = namesToIds.currentMapping(table);
+            if (identifier.isPresent() && !identifier.get().isPending()) {
+                return identifier.get().identifier();
+            }
+            log.info("Assigning table {} an identifier", LoggingArgs.tableRef(table));
+            SweepTableIdentifier afterPendingPut = namesToIds.storeAsPending(table, idToNames.getNextId());
```

Comment:
This reads a bit gross if we have to go through the loop more than once. This cas will fail and return the most recently written ID, which was inserted by the other `storeAsPending` call. My first thought was that the logic was incorrect, but since it doubles as a read, it works.

Reply:
I made it more clear. We certainly have to pick between a PUE or a CAS here